File: frontend/src/api/options.ts
import apiClient, { unwrapResponse } from './client';
import type { ApiResponse } from '../types';

export interface OptionPayload {
  ticker: string;
  trade_type: 'sell_put' | 'sell_call' | 'buy_call' | 'buy_put';
  strike_price: number;
  premium: number;
  contracts: number;
  open_date: string;
  expiry_date: string;
  status: 'open' | 'closed' | 'expired_worthless' | 'assigned';
  close_price?: number | null;
  notes?: string | null;
  broker?: string | null;
}

export async function getOptions() {
  const { data } = await apiClient.get<ApiResponse<any[]>>('/api/options');
  return unwrapResponse(data);
}

export async function getOption(id: number) {
  const { data } = await apiClient.get<ApiResponse<any>>(`/api/options/${id}`);
  return unwrapResponse(data);
}

export async function createOption(payload: OptionPayload) {
  const { data } = await apiClient.post<ApiResponse<any>>('/api/options', payload);
  return unwrapResponse(data);
}

export async function updateOption(id: number, payload: OptionPayload) {
  const { data } = await apiClient.put<ApiResponse<any>>(`/api/options/${id}`, payload);
  return unwrapResponse(data);
}

export async function deleteOption(id: number) {
  const { data } = await apiClient.delete<ApiResponse<null>>(`/api/options/${id}`);
  return unwrapResponse(data);
}

export async function getOptionsSummary() {
  const { data } = await apiClient.get<ApiResponse<any>>('/api/options/summary');
  return unwrapResponse(data);
}

export interface OptionQuote {
  bid: number | null;
  ask: number | null;
  last_price: number | null;
  current_price: number | null;
  unrealized_pnl: number | null;
}

export async function getOpenTradeQuotes() {
  const { data } =
    await apiClient.get<ApiResponse<Record<string, OptionQuote>>>('/api/options/quotes');
  return unwrapResponse(data);
}

File: backend/app/schemas/options.py
"""Pydantic schemas for options trades."""
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class OptionsTradeCreate(BaseModel):
    ticker: str = Field(..., min_length=1, max_length=5, pattern=r"^[A-Z]{1,5}$")
    trade_type: Literal["sell_put", "sell_call", "buy_call", "buy_put"]
    strike_price: float = Field(..., ge=0)
    premium: float = Field(..., ge=0)
    contracts: int = Field(..., ge=1)
    open_date: date
    expiry_date: date
    status: Literal["open", "closed", "expired_worthless", "assigned"]
    close_price: float | None = None
    notes: str | None = None
    broker: str | None = None


class OptionsTradeResponse(BaseModel):
    id: int
    ticker: str
    trade_type: str
    strike_price: float
    premium: float
    contracts: int
    open_date: date
    expiry_date: date
    status: str
    close_price: float | None
    notes: str | None
    broker: str | None = None
    created_at: datetime
    updated_at: datetime
    pnl: float | None = None

    model_config = {"from_attributes": True}


class OptionsSummary(BaseModel):
    total_pnl: float
    open_positions: int
    expiring_this_week: int
    win_rate: float

    model_config = {"from_attributes": True}


class OptionQuoteResponse(BaseModel):
    """Live market quote for a single open options trade, sourced from yfinance."""

    bid: float | None
    ask: float | None
    last_price: float | None
    current_price: float | None  # bid/ask midpoint, falls back to last_price
    unrealized_pnl: float | None
